core/services/relay/evm/read/multieventtype.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+)
+
+type EventQuery struct {
+	Filter           query.KeyFilter
+	EventBinding     *EventBinding
+	SequenceDataType any
+	IsValuePtr       bool
+	Address          common.Address
+}
+
+func MultiEventTypeQuery(ctx context.Context, lp logpoller.LogPoller, eventQueries []EventQuery, limitAndSort query.LimitAndSort) (sequences []commontypes.SequenceWithKey, err error) {
+	defer func() {
+		if err != nil {
+			if len(eventQueries) > 0 {
+				// TODO need to figure out what would be a sensible way to capture the error for all the queries
+				eq := eventQueries[0]
+
+				err = newErrorFromCall(err, Call{
+					ContractAddress: eq.Address,
+					ContractName:    eq.EventBinding.contractName,
+					ReadName:        eq.EventBinding.eventName,
+					ReturnVal:       eq.SequenceDataType,
+				}, "", eventReadType)
+			} else {
+				err = fmt.Errorf("no event queries provided: %w", err)
+			}
+		}
+	}()
+
+	for _, eq := range eventQueries {
+		if err = eq.EventBinding.validateBound(eq.Address); err != nil {
+			return nil, err
+		}
+	}
+
+	allFilterExpressions := make([]query.Expression, 0, len(eventQueries))
+	for idx, eq := range eventQueries {
+		var expressions []query.Expression
+
+		defaultExpressions := []query.Expression{
+			logpoller.NewAddressFilter(eq.Address),
+			logpoller.NewEventSigFilter(eq.EventBinding.hash),
+		}
+		expressions = append(expressions, defaultExpressions...)
+
+		remapped, remapErr := eq.EventBinding.remap(eq.Filter)
+		if remapErr != nil {
+			return nil, fmt.Errorf("error remapping filter: %w", err)
+		}
+		expressions = append(expressions, remapped.Expressions...)
+
+		filterExpression := query.And(expressions...)
+
+		allFilterExpressions[idx] = filterExpression
+	}
+
+	eventQuery := query.Or(allFilterExpressions...)
+
+	queryName := ""
+	for _, eq := range eventQueries {
+		queryName += eq.EventBinding.contractName + "-" + eq.Address.String() + "-" + eq.EventBinding.eventName + "-"
+	}
+	queryName = strings.TrimSuffix(queryName, "-")
+
+	logs, err := lp.FilteredLogs(ctx, []query.Expression{eventQuery}, limitAndSort, queryName)
+	if err != nil {
+		return nil, wrapInternalErr(err)
+	}
+
+	// no need to return an error. an empty list is fine
+	if len(logs) == 0 {
+		return []commontypes.SequenceWithKey{}, nil
+	}
+
+	sequences, err = decodeMultiEventTypeLogsIntoSequences(ctx, logs, eventQueries)
+	if err != nil {
+		return nil, err
+	}
+
+	return sequences, nil
+}
+
+func decodeMultiEventTypeLogsIntoSequences(ctx context.Context, logs []logpoller.Log, eventQueries []EventQuery) ([]commontypes.SequenceWithKey, error) {
+	sequences := make([]commontypes.SequenceWithKey, len(logs))
+
+	eventSigToEventQuery := map[common.Hash]EventQuery{}
+	for _, eq := range eventQueries {
+		eventSigToEventQuery[eq.EventBinding.hash] = eq
+	}
+
+	for idx := range logs {
```

Comment:
maybe?
```suggestion
	for idx, log := range logs {
```
then we don't need `eventSignatureHash` var